### PR TITLE
Guided filter scalability improvement

### PR DESCRIPTION
**Pull request by ralfbrown**
Flipped around the parallelization (parallel-process tiles one at a time instead of single-thread processing of multiple tiles in parallel), which improves the scalability when the filter radius is large and thus there are few tiles.  This also reduces memory requirements because only one tile is active at a time instead of multiple tiles.

Merged two of the four-channel and the remaining single-channel box filter into a 9-channel filter, which got a noticeable speedup despite no longer nicely matching up to vector instructions.  Memory bandwidth is still the limitation at high thread counts.

Timings on mire1:
Exposure+Blender feathering (test 0025)
Thr  / Before / After
1 / 4.176 / 3.501
2 / 2.514 / 1.920
4 / 1.329 / 1.072
6 / 1.126 / 0.791
8 / 1.110 / 0.636
12 / 0.762 / 0.524
16 / 0.741 / 0.461
24 / 0.719 / 0.398
32 / 0.727 / 0.386

Haze Removal
Thr / Before / After
1 / 1.746 / 1.529
2 / 0.998 / 0.877
4 / 0.566 / 0.504
6 / 0.408 / 0.364
8 / 0.333 / 0.316
12 / 0.255 / 0.229
16 / 0.214 / 0.202
24 / 0.185 / 0.176
32 / 0.188 / 0.187
(The guided filter accounts for less than half the runtime of the haze removal.)

### CODE REVIEW

File: src/common/guided_filter.c [hunk 1]
```diff
@@ -55,6 +55,18 @@
 # define _mm_prefetch(where,hint)
 #endif
 
+#ifndef dt_omp_shared
+#ifdef _OPENMP
+#if defined(__clang__) || __GNUC__ > 8
+# define dt_omp_shared(var, ...)  shared(var, __VA_ARGS__)
```
Comment (TurboGit) [inline src/common/guided_filter.c:61]: I suppose this should be:
`# define dt_omp_shared(...)  shared(__VA_ARGS__)`

To allow for a single argument to be passed.

Comment (ralfbrown) [reply src/common/guided_filter.c]: You're right, though interestingly enough GCC9 is fine with that definition and a single arg.  I had just copied the definition I used in common/nlmeans_core.c; now that it's used in two places, it should probably be refactored into a header file (the obvious place being right after the definition of dt_omp_firstprivate in common/darktable.h).

**Comment by TurboGit** [comment]
Works for me! Thanks.

File: src/common/guided_filter.c (hunk 1 above)
Comment (TurboGit) [reply src/common/guided_filter.c]: I'll try to do that indeed.

**Pull request merged**